File: src/stores/notes.ts
import { defineStore } from 'pinia'
import { invoke } from '@/lib/tauri'
import { ref } from 'vue'

/**
 * Global quick-notes scratchpad, persisted in the app SQLite DB (notes table)
 * via Tauri commands. Each note has a short title + markdown content and may
 * optionally be linked to a project (`project_id`). Array order mirrors the
 * `position` column — drag-and-drop in the view reorders locally then persists
 * the new id order.
 *
 * Mutations apply optimistically for a snappy UI and refetch on error to resync.
 */
export interface Note {
  id: string
  title: string
  content: string
  project_id: string | null
  /** Capture context: the AI run it came from, for the backlink. */
  run_id: string | null
  position: number
  created_at: string
  updated_at: string
}

export const useNotesStore = defineStore('notes', () => {
  const notes = ref<Note[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function fetchNotes(projectId?: string | null) {
    loading.value = true
    error.value = null
    try {
      notes.value = await invoke<Note[]>('list_notes', { projectId: projectId ?? null })
    } catch (e) {
      error.value = String(e)
    } finally {
      loading.value = false
    }
  }

  async function add(
    title: string,
    content: string,
    projectId?: string | null,
    runId?: string | null,
  ) {
    const t = title.trim()
    const c = content.trim()
    if (!t && !c) return
    try {
      const note = await invoke<Note>('add_note', {
        title: t,
        content: c,
        projectId: projectId ?? null,
        // A run backlink only means anything alongside its project.
        runId: projectId ? runId ?? null : null,
      })
      notes.value.unshift(note)
      return note
    } catch (e) {
      error.value = String(e)
      await fetchNotes()
    }
  }

  async function update(id: string, title: string, content: string) {
    const t = title.trim()
    const c = content.trim()
    if (!t && !c) return
    const n = notes.value.find(n => n.id === id)
    const prev = n ? { title: n.title, content: n.content } : null
    if (n) {
      n.title = t
      n.content = c // optimistic
    }
    try {
      await invoke('update_note', { id, title: t, content: c })
      if (n) n.updated_at = new Date().toISOString()
    } catch (e) {
      error.value = String(e)
      if (n && prev) {
        n.title = prev.title
        n.content = prev.content
      }
    }
  }

  async function setProject(id: string, projectId: string | null) {
    const n = notes.value.find(n => n.id === id)
    const prev = n ? { projectId: n.project_id, runId: n.run_id } : null
    if (n) {
      n.project_id = projectId // optimistic
      if (!projectId) n.run_id = null // a run backlink can't outlive its project link
    }
    try {
      await invoke('set_note_project', { id, projectId })
    } catch (e) {
      error.value = String(e)
      if (n && prev) {
        n.project_id = prev.projectId
        n.run_id = prev.runId
      }
    }
  }

  async function remove(id: string) {
    const i = notes.value.findIndex(n => n.id === id)
    if (i === -1) return
    const [removed] = notes.value.splice(i, 1) // optimistic
    try {
      await invoke('delete_note', { id })
    } catch (e) {
      error.value = String(e)
      notes.value.splice(i, 0, removed)
    }
  }

  /** Bulk-delete notes by id (optimistic), refetching to resync on error. */
  async function removeMany(ids: string[]) {
    if (ids.length === 0) return
    const target = new Set(ids)
    const snapshot = notes.value.slice()
    notes.value = notes.value.filter(n => !target.has(n.id)) // optimistic
    try {
      await invoke('delete_notes', { ids })
    } catch (e) {
      error.value = String(e)
      notes.value = snapshot
    }
  }

  /** Move the note at `from` to `to` (drag-and-drop reorder), then persist. */
  async function reorder(from: number, to: number) {
    if (from === to) return
    if (from < 0 || from >= notes.value.length) return
    if (to < 0 || to >= notes.value.length) return
    const snapshot = notes.value.slice()
    const [moved] = notes.value.splice(from, 1)
    notes.value.splice(to, 0, moved)
    try {
      await invoke('reorder_notes', { ids: notes.value.map(n => n.id) })
    } catch (e) {
      error.value = String(e)
      notes.value = snapshot
    }
  }

  return { notes, loading, error, fetchNotes, add, update, setProject, remove, removeMany, reorder }
})

File: src-tauri/src/commands/notes.rs
use crate::db::Db;
use serde::Serialize;
use sqlx::Row;
use tauri::State;
use uuid::Uuid;

/// A single quick note. `title` is a short label (may be empty — the UI falls
/// back to the first line of `content`), `content` is markdown. `project_id`
/// optionally links the note to a project, and `run_id` to the AI run it was
/// captured from (so the UI can offer a backlink). `position` orders the list
/// (ascending: lower value sits higher / top). Drag-and-drop reorder rewrites
/// positions.
#[derive(Debug, Serialize, Clone)]
pub struct Note {
    pub id: String,
    pub title: String,
    pub content: String,
    pub project_id: Option<String>,
    pub run_id: Option<String>,
    pub position: i64,
    pub created_at: String,
    pub updated_at: String,
}

fn row_to_note(row: &sqlx::sqlite::SqliteRow) -> Note {
    Note {
        id: row.get("id"),
        title: row.get("title"),
        content: row.get("content"),
        project_id: row.get("project_id"),
        run_id: row.get("run_id"),
        position: row.get("position"),
        created_at: row.get("created_at"),
        updated_at: row.get("updated_at"),
    }
}

/// List notes, newest-priority first. When `project_id` is `Some`, only notes
/// linked to that project are returned; otherwise all notes are returned.
#[tauri::command]
pub async fn list_notes(
    db: State<'_, Db>,
    project_id: Option<String>,
) -> Result<Vec<Note>, String> {
    let rows = if let Some(pid) = project_id {
        sqlx::query(
            "SELECT id, title, content, project_id, run_id, position, created_at, updated_at \
             FROM notes WHERE project_id = ? \
             ORDER BY position ASC, updated_at DESC",
        )
        .bind(pid)
        .fetch_all(db.inner())
        .await
        .map_err(|e| e.to_string())?
    } else {
        sqlx::query(
            "SELECT id, title, content, project_id, run_id, position, created_at, updated_at \
             FROM notes ORDER BY position ASC, updated_at DESC",
        )
        .fetch_all(db.inner())
        .await
        .map_err(|e| e.to_string())?
    };
    Ok(rows.iter().map(row_to_note).collect())
}

#[tauri::command]
pub async fn add_note(
    db: State<'_, Db>,
    title: String,
    content: String,
    project_id: Option<String>,
    run_id: Option<String>,
) -> Result<Note, String> {
    let title = title.trim().to_string();
    let content = content.trim().to_string();
    if title.is_empty() && content.is_empty() {
        return Err("a title or content is required".into());
    }
    let id = Uuid::new_v4().to_string();
    let now = chrono::Utc::now().to_rfc3339();
    // New notes go to the top (highest priority) by taking a position below the
    // current minimum.
    let min_pos: Option<i64> = sqlx::query("SELECT MIN(position) AS min_pos FROM notes")
        .fetch_one(db.inner())
        .await
        .map_err(|e| e.to_string())?
        .get("min_pos");
    let position = min_pos.unwrap_or(0) - 1;

    sqlx::query(
        "INSERT INTO notes (id, title, content, project_id, run_id, position, created_at, updated_at) \
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
    )
    .bind(&id)
    .bind(&title)
    .bind(&content)
    .bind(&project_id)
    .bind(&run_id)
    .bind(position)
    .bind(&now)
    .bind(&now)
    .execute(db.inner())
    .await
    .map_err(|e| e.to_string())?;

    Ok(Note {
        id,
        title,
        content,
        project_id,
        run_id,
        position,
        created_at: now.clone(),
        updated_at: now,
    })
}

#[tauri::command]
pub async fn update_note(
    db: State<'_, Db>,
    id: String,
    title: String,
    content: String,
) -> Result<(), String> {
    let title = title.trim().to_string();
    let content = content.trim().to_string();
    if title.is_empty() && content.is_empty() {
        return Err("a title or content is required".into());
    }
    let now = chrono::Utc::now().to_rfc3339();
    sqlx::query("UPDATE notes SET title = ?, content = ?, updated_at = ? WHERE id = ?")
        .bind(&title)
        .bind(&content)
        .bind(&now)
        .bind(&id)
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

/// Link or unlink a note from a project. Pass `None` to clear the link; the run
/// backlink is cleared with it, since a run only makes sense inside its project.
#[tauri::command]
pub async fn set_note_project(
    db: State<'_, Db>,
    id: String,
    project_id: Option<String>,
) -> Result<(), String> {
    let now = chrono::Utc::now().to_rfc3339();
    if project_id.is_none() {
        sqlx::query(
            "UPDATE notes SET project_id = NULL, run_id = NULL, updated_at = ? WHERE id = ?",
        )
        .bind(&now)
        .bind(&id)
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
        return Ok(());
    }
    sqlx::query("UPDATE notes SET project_id = ?, updated_at = ? WHERE id = ?")
        .bind(&project_id)
        .bind(&now)
        .bind(&id)
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub async fn delete_note(db: State<'_, Db>, id: String) -> Result<(), String> {
    sqlx::query("DELETE FROM notes WHERE id = ?")
        .bind(&id)
        .execute(db.inner())
        .await
        .map_err(|e| e.to_string())?;
    Ok(())
}

/// Bulk-delete notes by id in a single transaction, so a failure mid-way rolls
/// back the whole batch instead of leaving a partial deletion.
#[tauri::command]
pub async fn delete_notes(db: State<'_, Db>, ids: Vec<String>) -> Result<(), String> {
    if ids.is_empty() {
        return Ok(());
    }
    let mut tx = db.inner().begin().await.map_err(|e| e.to_string())?;
    for id in &ids {
        sqlx::query("DELETE FROM notes WHERE id = ?")
            .bind(id)
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;
    }
    tx.commit().await.map_err(|e| e.to_string())?;
    Ok(())
}

/// Persist a new order. The frontend sends the full list of ids top-to-bottom;
/// each row's position becomes its index, so the ordering is stable regardless
/// of prior position values. Done in one transaction to avoid partial reorders.
#[tauri::command]
pub async fn reorder_notes(db: State<'_, Db>, ids: Vec<String>) -> Result<(), String> {
    let mut tx = db.inner().begin().await.map_err(|e| e.to_string())?;
    for (index, id) in ids.iter().enumerate() {
        sqlx::query("UPDATE notes SET position = ? WHERE id = ?")
            .bind(index as i64)
            .bind(id)
            .execute(&mut *tx)
            .await
            .map_err(|e| e.to_string())?;
    }
    tx.commit().await.map_err(|e| e.to_string())?;
    Ok(())
}
